File: lib/data/models/waste_report.dart
import 'json_helpers.dart';

class WasteReport {
  final int reportId;
  final int submittedBy;
  final String submittedByName;
  final List<int> wasteTypeIds;
  final List<String> wasteTypeNames;
  final String imageUrl;
  final String description;
  final double? latitude;
  final double? longitude;
  final String status;
  final DateTime? createdAt;

  const WasteReport({
    this.reportId = 0,
    this.submittedBy = 0,
    this.submittedByName = '',
    this.wasteTypeIds = const [],
    this.wasteTypeNames = const [],
    this.imageUrl = '',
    this.description = '',
    this.latitude,
    this.longitude,
    this.status = '',
    this.createdAt,
  });

  factory WasteReport.fromJson(Map<String, dynamic> json) {
    return WasteReport(
      reportId: JsonHelpers.intValue(json, [
        'reportId',
        'ReportId',
        'id',
        'Id',
      ]),
      submittedBy: JsonHelpers.intValue(json, [
        'submittedBy',
        'SubmittedBy',
        'userId',
        'UserId',
      ]),
      submittedByName: JsonHelpers.stringValue(json, [
        'submittedByName',
        'SubmittedByName',
      ]),
      wasteTypeIds: JsonHelpers.intList(json, ['wasteTypeIds', 'WasteTypeIds']),
      wasteTypeNames: JsonHelpers.stringList(json, [
        'wasteTypeNames',
        'WasteTypeNames',
        'wasteTypeName',
        'WasteTypeName',
      ]),
      imageUrl: JsonHelpers.stringValue(json, [
        'imageUrl',
        'ImageUrl',
        'reportImageUrl',
        'ReportImageUrl',
      ]),
      description: JsonHelpers.stringValue(json, [
        'description',
        'Description',
      ]),
      latitude: JsonHelpers.doubleValue(json, ['latitude', 'Latitude']),
      longitude: JsonHelpers.doubleValue(json, ['longitude', 'Longitude']),
      status: JsonHelpers.stringValue(json, ['status', 'Status']),
      createdAt: JsonHelpers.dateValue(json, ['createdAt', 'CreatedAt']),
    );
  }

  bool get canCancel => status == 'Pending';
  bool get canFeedback => status == 'Completed' || status == 'Collected';
}

File: src/Services/WasteReportService/WasteReportService.Domain/Entities/WasteReport.cs
namespace WasteReportService.Domain.Entities;

public sealed class WasteReport
{
    public int ReportId { get; set; }
    public int SubmittedBy { get; set; }
    public string ImageUrl { get; set; } = null!;
    public decimal Latitude { get; set; }
    public decimal Longitude { get; set; }
    public string? Description { get; set; }
    public string? Status { get; set; }
    public string? EstimatedSize { get; set; }
    public DateTime? CreatedAt { get; set; }
    public int? DistrictId { get; set; }
    public District? District { get; set; }
    public ICollection<AiWastePrediction> AiWastePredictions { get; set; } = new List<AiWastePrediction>();
    public ICollection<WasteType> WasteTypes { get; set; } = new List<WasteType>();
}
